parachain-system/src/lib.rs

```diff
@@ -96,6 +106,19 @@ decl_storage! {
 		/// This data is also absent from the genesis.
 		HostConfiguration get(fn host_configuration): Option<AbridgedHostConfiguration>;
 
+		/// The last downward message queue chain head we have observed or all zeroes if we haven't observed
+		/// any.
+		///
+		/// This value is loaded before and saved after processing inbound downward messages carried
+		/// by the system inherent.
+		LastDmqMqcHead: relay_chain::Hash;
+		/// The message queue chain heads we have observed per each channel incoming channel. If the
+		/// value is missing it is treated as all zeroes.
```

Comment:
The reason why a missing value is treated with all zeroes is because we state that on the polkadot side, [here](https://github.com/paritytech/polkadot/blob/1dd53d4345b13b15f67d9c4f54d8b63cc0ba1902/runtime/parachains/src/hrmp.rs#L78-L85)

It says to us that the previous message queue chain head for an empty queue is all zeros. So whenever we get an inherent that has some MQCs that we previously didn't know about, we treat them as all zeros.

However, after processing the batch we overwrite the whole structure with the MQC heads that we observed in the last round.

Does this make sense?

Reply:
Yeah that also makes sense from the code, however isn't it this case that this storage item will never have the empty head? Because when we insert some value for a parachain, it will always be a non-all-zero head?

Reply:
Actually, the values may have an empty head. This can happen when we observe a new MQC head from persisted validation data but receive no messages. In this case a zero will end up in this storage item.